src/test/regress/spec/isolation_rebalancer_deferred_drop.spec

```diff
@@ -0,0 +1,66 @@
+// we use 15 as the partition key value through out the test
+// so setting the corresponding shard here is useful
+setup
+{
+  SELECT citus_internal.replace_isolation_tester_func();
+  SELECT citus_internal.refresh_isolation_tester_prepared_statement();
+
+CREATE OR REPLACE FUNCTION master_defer_delete_shards()
+    RETURNS int
+    LANGUAGE C STRICT
+    AS 'citus', $$master_defer_delete_shards$$;
+COMMENT ON FUNCTION master_defer_delete_shards()
+    IS 'remove orphaned shards';
+
+	SET citus.shard_count TO 8;
+	SET citus.shard_replication_factor TO 1;
+    SET citus.defer_drop_after_shard_move TO ON;
+	CREATE TABLE t1 (x int PRIMARY KEY, y int);
+	SELECT create_distributed_table('t1', 'x');
+
+	SELECT get_shard_id_for_distribution_column('t1', 15) INTO selected_shard;
+}
+
+teardown
+{
+  SELECT citus_internal.restore_isolation_tester_func();
+
+  DROP TABLE selected_shard;
+  DROP TABLE t1;
+}
+
+
+session "s1"
+
+step "s1-begin"
+{
+    BEGIN;
+}
+
+step "s1-move-placement"
+{
+        SET citus.defer_drop_after_shard_move TO ON;
+    	SELECT master_move_shard_placement((SELECT * FROM selected_shard), 'localhost', 57637, 'localhost', 57638);
+}
+
+step "s1-drop-marked-shards"
+{
+    SELECT public.master_defer_delete_shards();
+}
+
+step "s1-commit"
+{
+    COMMIT;
+}
+
+session "s2"
+
+step "s2-drop-marked-shards"
+{
+    SELECT public.master_defer_delete_shards();
+}
+
+permutation "s1-begin" "s1-move-placement" "s1-drop-marked-shards" "s2-drop-marked-shards" "s1-commit"
```

Comment:
The transactional semantics are a quite confusing here. It's probably a good thing users cannot actually call master_defer_delete_shards.

Reply:
I was thinking of making this master_defer_delete_shards public because after making the deferred behavior the default, we will also need to think about a default for `defer_shard_delete_interval`. I have two questions/comments here:
- What should be the default value for this, currently it is -1, so disabled but it sure needs to be "enabled" and if this value if high then users might be confused (as in they are possibly rebalancing to create some disk space but they end up having more since the old placements are still there. At a high level, why would users want to set this value? I guess they would want the old placements to be gone as soon as possible as they are "garbage" at that point.
- If users don't want to wait, we will probably suggest to set this setting(defer_shard_delete_interval) to 0 and then retry. So it seems simpler to call master_defer_delete_shards in this case, would that make sense to expose this for operational convenience.

From user point of view, the simplest thing seems to be dropping the old placements when we can (as in when we can take the locks)